borgmatic/commands/completion.py

```diff
@@ -55,3 +58,55 @@ def bash_completion():
             '\ncomplete -o bashdefault -o default -F complete_borgmatic borgmatic',
         )
     )
+
+def build_fish_flags(action: Action):
+    '''
+    Given an argparse.Action instance, return a string containing the fish flags for that action.
+    '''
+    if action.metavar and action.metavar == 'PATH' or action.metavar == 'FILENAME':
+        return '-r -F'
+    else:
+        return '-f'
+
+def fish_completion():
+    '''
+    Return a fish completion script for the borgmatic command. Produce this by introspecting
+    borgmatic's command-line argument parsers.
+    '''
+    top_level_parser, subparsers = arguments.make_parsers()
+
+    all_subparsers = ' '.join(action for action in subparsers.choices.keys())
+
+    # Avert your eyes.
+    return '\n'.join(
+        (
+            'function __borgmatic_check_version',
+            '    set this_filename (status current-filename)',
+            '    set this_script (cat $this_filename 2> /dev/null)',
+            '    set installed_script (borgmatic --fish-completion 2> /dev/null)',
+            '    if [ "$this_script" != "$installed_script" ] && [ "$installed_script" != "" ]',
+            '        echo "{}"'.format(upgrade_message('fish', 'borgmatic --fish-completion | sudo tee $this_filename', '$this_filename')),
+            '    end',
+            'end',
+            '__borgmatic_check_version &',
+        ) + (
+            '\n# subparser completions',
+        ) + tuple(
+            '''complete -c borgmatic -a '%s' -d %s -f -n "not __fish_seen_subcommand_from %s"'''
+            % (actionStr, shlex.quote(subparser.description), all_subparsers)
+            for actionStr, subparser in subparsers.choices.items()
+        ) + (
+            '\n# global flags',
+        ) + tuple(
+            '''complete -c borgmatic -a '%s' -d %s %s'''
+            % (' '.join(action.option_strings), shlex.quote(action.help), build_fish_flags(action))
+            for action in top_level_parser._actions
+        ) + (
+            '\n# subparser flags',
+        ) + tuple(
+            '''complete -c borgmatic -a '%s' -d %s -n "__fish_seen_subcommand_from %s" %s'''
+            % (' '.join(action.option_strings), shlex.quote(action.help), actionStr, build_fish_flags(action))
+            for actionStr, subparser in subparsers.choices.items()
+            for action in subparser._actions
+        )
+    )
```

Comment:
yep! its not quite as good as I would like it to be, but I wasn't able to combine it to group equivalent arguments. Fish doesn't seem to really appreciate the "everything is actually an argument" approach to argparsing, it really wants them to be more like flags associated with args.

However, it has the correct arguments with descriptions, and only shows the contextual args contextually, so that's pretty decent.
![scr_13:11:10](https://user-images.githubusercontent.com/72410860/236318599-09844544-421f-4639-b075-8de2c526ee3f.png)
![image](https://user-images.githubusercontent.com/72410860/236318548-10054fbe-31fd-41ed-9788-063feca4360d.png)

In particular, knowing when file completions are correct seems to be a bit busted...
![image](https://user-images.githubusercontent.com/72410860/236318831-bbfb1eb5-d462-4dd0-a6f5-f2c76774850b.png)
I can try and fix that, it seems like the conditional arg gets associated in some way.

Reply:
Pretty cool! Much nicer than the Bash completions, at least the way it's presented to the user. It's up to you whether it's worth sinking more time into the file completions or not.